### PR TITLE
feat: Support for `ProximitySensor` on Android

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): closes https://github.com/unoplatform/uno/issues/12304

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

- Not supported

## What is the new behavior?

- Supported on Android

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/SamplesApp/SamplesApp.netcoremobile/Android/AssemblyInfo.Android.cs [hunk 1]
```diff
@@ -13,3 +13,4 @@
 [assembly: UsesPermission("android.permission.RECORD_AUDIO")]
 [assembly: UsesPermission("android.permission.CAMERA")]
 [assembly: UsesPermission("android.permission.WRITE_EXTERNAL_STORAGE")]
+[assembly: UsesPermission("android.hardware.sensor.proximity")]
```
Comment (Youssef1313) [inline src/SamplesApp/SamplesApp.netcoremobile/Android/AssemblyInfo.Android.cs:16]: Not important, but there is another AssemblyInfo file for the Xamarin-based project.

File: src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#pragma warning disable CS0067
+	public event EventHandler<DeviceInformation>? WatchAdded;
+
+	public event EventHandler<DeviceInformation>? WatchEnumerationCompleted;
+
+	public event EventHandler<DeviceInformationUpdate>? WatchRemoved;
+
+	public event EventHandler<object>? WatchStopped;
+
+	public event EventHandler<DeviceInformationUpdate>? WatchUpdated;
```
Comment (Youssef1313) [inline src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.cs:26]: Should they be invoked properly?

File: src/Uno.UWP/Devices/Sensors/ProximitySensor.Android.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	partial void StopReading()
+	{
+		if (_listener is not null)
+		{
+			SensorHelpers.GetSensorManager().UnregisterListener(_listener, _sensor);
+			_listener.Dispose();
+			_listener = null;
+		}
+	}
```
Comment (Youssef1313) [inline src/Uno.UWP/Devices/Sensors/ProximitySensor.Android.cs:68]: Quite curious about thread-safety here.

Is it realistic that two threads call `StopReading`, and both go into the if condition, but one of them will set _listener to null earlier, then the other will throw NullReferenceException?

Comment (Youssef1313) [reply src/Uno.UWP/Devices/Sensors/ProximitySensor.Android.cs:68]: Realized now that the implementation of StartStopEventWrapper is already thread safe.
 https://github.com/unoplatform/uno/blob/0bf7c32d04800952fa732298268a7a3f1c548c8b/src/Uno.UWP/Helpers/StartStopEventWrapper.cs#L59-L71

**Review by jeromelaban — APPROVED** [review]

**Review by ajpinedam — APPROVED** [review]

**Comment by jeromelaban** [comment]
```
ProximitySensorDeviceClassProvider.cs(7,7): Error CS0246: The type or namespace name 'Android' could not be found (are you missing a using directive or an assembly reference?)
```

File: src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.cs (hunk 2 above)
Comment (MartinZikmund) [reply src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.cs]: The watcher is marked as not supported, so they are not used. I agree that the API should be improved later to separate interface though.

File: src/SamplesApp/SamplesApp.netcoremobile/Android/AssemblyInfo.Android.cs (hunk 1 above)
Comment (MartinZikmund) [reply src/SamplesApp/SamplesApp.netcoremobile/Android/AssemblyInfo.Android.cs]: Actually turns out it is not needed, so I removed it here

File: src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.Android.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+		}
+
+		foreach (var sensor in sensors ?? Array.Empty<Sensor>())
+		{
+			var deviceIdentifier = new DeviceIdentifier(sensor.Id.ToString(CultureInfo.InvariantCulture), DeviceClassGuids.ProximitySensor);
+			var deviceInformation = new DeviceInformation(deviceIdentifier);
+			devices.Add(deviceInformation);
+		}
+
+		return Task.FromResult(devices.ToArray());
```
Comment (jeromelaban) [inline src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.Android.cs:55]: I wonder, it's probably useful to run most of this on a background thread

File: src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.Android.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	public event EventHandler<DeviceInformation>? WatchEnumerationCompleted;
+
+	public event EventHandler<DeviceInformationUpdate>? WatchRemoved;
+
+	public event EventHandler<object>? WatchStopped;
+
+	public event EventHandler<DeviceInformationUpdate>? WatchUpdated;
+#pragma warning restore CS0067
+
+	public Task<DeviceInformation[]> FindAllAsync() => Task.Run(() =>
```
Comment (Youssef1313) [inline src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.Android.cs:29]: Why do we need Task.Run?

Comment (MartinZikmund) [reply src/Uno.UWP/Devices/Enumeration/Internal/Providers/ProximitySensor/ProximitySensorDeviceClassProvider.Android.cs:29]: It was suggested by Jerome above, it is not strictly necessary, but won't hurt, as the method is `Task` based anyway

**Pull request merged**